cmake/Modules/CodeCoverage.cmake

```diff
@@ -139,6 +139,81 @@ function(ADD_CODE_COVERAGE)
+    string(REPLACE ";" "," Coverage_EXCLUDES_STR "${Coverage_EXCLUDES}")
+    set(OMIT_FLAGS "--omit=\"${Coverage_EXCLUDES_STR}\"")
+    set(INCLUDE_FLAGS "--include=\"${REAL_SOURCE_DIR}/*\"")
+
+    # add target for non-test repo
+    # we need to set run_tests_${PROJECT_NAME} and _run_tests_${PROJECT_NAME} for repo with no-test
+    # otherwise test fails.
+    # catkin tools executed either runrun_tests_${PROJECT_NAME} or _run_tests_${PROJECT_NAME}.
+    # run_tests_${PROJECT_NAME} and _run_tests_${PROJECT_NAME} does not run simaltaneously. 
+    # _run_tests_${PROJECT_NAME} is for cleaning test result and run tests.
+    # https://github.com/ros/catkin/commit/f931db5c8c14475a9d74ffc65b9dbbe45c98d11d
+    if(NOT TARGET run_tests_${PROJECT_NAME})
+      add_custom_target(run_tests_${PROJECT_NAME})
+    endif()
+
+    if(NOT TARGET _run_tests_${PROJECT_NAME})
+      # create hidden meta target which depends on hidden test targets which depend on clean_test_results
+      add_custom_target(_run_tests_${PROJECT_NAME})
+      # run_tests depends on this hidden target hierarchy to clear test results before running all tests
+     add_dependencies(run_tests _run_tests_${PROJECT_NAME})
+    endif()
+
+    if(NOT DEFINED CATKIN_ENABLE_TESTING OR CATKIN_ENABLE_TESTING)
+      # create python base coverage directory
+      add_custom_target(
+        create_python_base_coverage_dir "${CMAKE_COMMAND}" "-E" "make_directory" ${PROJECT_BINARY_DIR}/python_base_coverage
+      )
+
+      # find code_coverage path
+      if(code_coverage_SOURCE_DIR)
+        set(_code_coverage_SOURCE_DIR ${code_coverage_SOURCE_DIR})
+      elseif(code_coverage_SOURCE_PREFIX)
+        set(_code_coverage_SOURCE_DIR ${code_coverage_SOURCE_PREFIX})
+      else(code_coverage_SOURCE_PREFIX)
+        set(_code_coverage_SOURCE_DIR ${code_coverage_PREFIX}/share/code_coverage)
+      endif()
+
+      # create python base coverage report
+      # generate_base_coverage.py list up python files in the repo and generate base coverage report
+      # base coverage report is needed to cover all python files, including non-tested files.
+      add_custom_target(run_tests_${PROJECT_NAME}_python_base_coverage_report
+        COMMAND ${_code_coverage_SOURCE_DIR}/scripts/generate_base_coverage.py ${PROJECT_SOURCE_DIR}
+                --output ${PROJECT_BINARY_DIR}/python_base_coverage
+        COMMAND ${PYTHON_COVERAGE_PATH} report ${INCLUDE_FLAGS} ${OMIT_FLAGS} || echo "WARNING: No python base report to output"
+        COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_base_python.xml ${INCLUDE_FLAGS} ${OMIT_FLAGS} || echo "WARNING: No base python xml to output"
+        COMMAND ${CMAKE_COMMAND} -E copy ${PROJECT_BINARY_DIR}/python_base_coverage/${Coverage_NAME}_base_python.xml ${PROJECT_BINARY_DIR}/ || echo "WARNING: No base python xml to copy"
+        DEPENDS create_python_base_coverage_dir
+                ${_code_coverage_SOURCE_DIR}/scripts/generate_base_coverage.py
```

Comment:
As far as I understand, this target should run again if a file is added to or removed from `python_directories` in `generate_base_coverage.py`.  Are these `DEPENDS` enough to satisfy this requirement?
Do we need to add the directories to `DEPENDS`?

Reply:
you are right.
I add directories which `generate_base_coverage.py` search for.
https://github.com/GITAI/code_coverage/pull/6/commits/c27dcd1ba8cfa30c59b308ab1e2dc348bd1a0938